docs/core/common/assets/js/api/core/ui-states.md

```diff
@@ -0,0 +1,178 @@
+				'off': this.onStateOff,
+			};
+		}
+ 
+		onChange( oldValue, newValue ) {
+			// Callback that runs on state change.
+			console.log( { oldValue, newValue } );
+		}
+	}
+	```
+
+  > Legend
+
+  | Name                          | Format - Description                                      | Example value
+    |-------------------------------|-----------------------------------------------------------|---------------------
+  |`{FILE_NAME_CAMEL_CASE}`       | CamelCase representation of the current file name.        | `DirectionMode`
+  |`{FILE_NAME_WITHOUT_EXTENSION}`| Current file name without the `.js` extension.            | `direction-mode`
+  |`{CUSTOM_PREFIX_KEBAB_CASE}`   | A kebab-case string.                                      | `some-custom-prefix`
+  |`{OPTION_VALUE_KEBAB_CASE}`    | A kebab-case string.                                      | `option-1`
+
+  > Example - A UI state for editing direction-mode:
+  
+	Let's assume we've created the following UI state: 
+  
+	```javascript
+	import UiStateBase from 'elementor-api/core/states/ui-state-base';
+	
+	// A good practice is to export the state options as constants so you 
+	// can use them when you set the state from another place.
+	export const DIRECTION_ROW = 'row';
+	export const DIRECTION_COLUMN = 'column';
+	
+	export class DirectionMode extends UiStateBase {
+		getId() {
+			return 'direction-mode';
+		}
+		
+		getPrefix() {
+			return 'document';
+		}
+		
+		getScopes() {
+			return [
+				window.document.body,
+				elementor.$previewContents[ 0 ],
+			];
+		}
+		
+		getOptions() {
+			return {
+				[ DIRECTION_ROW ]: '',
+				[ DIRECTION_COLUMN ]: '',
+			};
+		}
+		
+		onChange( oldValue, newValue ) {
+			console.log( 'Direction mode changed', { oldValue, newValue } );
+		}
+	}
+	```
+ 
+	Later on, on any state change, `onChange()` will be fired and will log the values to the console.
+  
+	In addition, a CSS class ( `e-ui-state--document-direction-mode__column` [ or `__row`, depends on the state value ] ) will be added to both of the scopes that we provided ( editor & preview in this case ).
+
+	Plus, a custom event ( `e-ui-state:document/direction-mode` ) will be fired to the scopes, and we can listen to that
+  	event using a simple `addEventListener()`:
+
+	```javascript
+	import { DIRECTION_ROW } from 'elementor-document/ui-states/direction-mode';
+ 
+	// Change the `direction-mode` state:
+	$e.uiStates.set( 'document/direction-mode', DIRECTION_ROW );
+ 
+ 
+	// Listen to changes:
+	window.document.body.addEventListener( 'e-ui-state:document/direction-mode', ( e ) => {
+		const { oldValue, newValue } = e.detail;
+		
+		console.log( 'Direction mode changed', { oldValue, newValue } );
+	} );
+
+	// OR
+
+	elementor.$previewContents[ 0 ].addEventListener( 'e-ui-state:document/direction-mode', ( e ) => {
+		const { oldValue, newValue } = e.detail;
+		
+		console.log( 'Direction mode changed', { oldValue, newValue } );
+	} );
+	```
+
+### [Back](../readme.md) 
```

Comment:
Not sure if it should be added, but, don't you think maybe to write an explanation of why this feature has been added (reasons and such) and what benefits does it provide?

Reply:
Good idea
Done.